packages/zero-client/src/client/delete-clients-manager.ts

```diff
@@ -0,0 +1,91 @@
+import type {LogContext} from '@rocicorp/logger';
+import type {Store} from '../../../replicache/src/dag/store.ts';
+import {
+  getDeletedClients,
+  setDeletedClients,
+} from '../../../replicache/src/deleted-clients.ts';
+import type {ClientID} from '../../../replicache/src/sync/ids.ts';
+import {
+  withRead,
+  withWrite,
+} from '../../../replicache/src/with-transactions.ts';
+import type {DeleteClientsMessage} from '../../../zero-protocol/src/delete-clients.ts';
+
+/**
+ * Replicache will tell us when it deletes clients from the persistent storage
+ * due to GC. When this happens we tell the server about the deleted clients. We
+ * also store the deleted clients in IDB in case the server is currently
+ * offline.
+ *
+ * The server will reply with the client it actually deleted. When we get that
+ * we remove those IDs from our local storage.
+ */
+export class DeleteClientsManager {
+  readonly #send: (msg: DeleteClientsMessage) => void;
+  readonly #lc: LogContext;
+  readonly #dagStore: Store;
+
+  constructor(
+    send: (msg: DeleteClientsMessage) => void,
+    dagStore: Store,
+    lc: LogContext,
+  ) {
+    this.#send = send;
+    this.#dagStore = dagStore;
+    this.#lc = lc;
+  }
+
+  /**
+   * This gets called by Replicache when it deletes clients from the persistent
+   * storage.
+   */
+  async onClientsDeleted(clientIDs: ClientID[]): Promise<void> {
+    this.#lc.debug?.('onClientsDeleted:', clientIDs);
+    const deletedClients = await withWrite(this.#dagStore, async dagWrite => {
```

Comment:
Maybe Replicache should be adding the deleted clients to the perdag deleted clients automatically at its layer, and then calling the onClientDeleted callback to inform layers above that things have been added.